### PR TITLE
gh-114115: Update documentation of array.array

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* Issue: gh-114115

----
📚 Documentation preview 📚: https://cpython-previews--114117.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/array.rst [hunk 1]
```diff
@@ -79,14 +79,16 @@ The module defines the following type:
 .. class:: array(typecode[, initializer])
 
    A new array whose items are restricted by *typecode*, and initialized
-   from the optional *initializer* value, which must be a list, a
-   :term:`bytes-like object`, or iterable over elements of the
-   appropriate type.
+   from the optional *initializer* value, which must be a :class:`bytes`
+   or :class:`bytearray` object, a unicode string, or iterable over elements
+   of the appropriate type.
```
Comment (serhiy-storchaka) [inline Doc/library/array.rst]: Only bytes and bytearray are special, bytes-like objects in general are not supported. List is a particular type of iterable.

File: Doc/library/array.rst [hunk 2]
```diff
@@ -79,14 +79,16 @@ The module defines the following type:
...
-   If given a list or string, the initializer is passed to the new array's
-   :meth:`fromlist`, :meth:`frombytes`, or :meth:`fromunicode` method (see below)
-   to add initial items to the array.  Otherwise, the iterable initializer is
-   passed to the :meth:`extend` method.
+   If given a :class:`bytes` or :class:`bytearray` object, the initializer
+   is passed to the new array's :meth:`frombytes` method;
+   if given a unicode string, the initializer is passed to the
+   :meth:`fromunicode` method;
+   otherwise, the initializer's iterator is passed to the :meth:`extend` method
+   to add initial items to the array.
```
Comment (serhiy-storchaka) [inline Doc/library/array.rst]: Only bytes, bytearray and str are special, other types are just iterables, although the constructor has some more special cases for optimization.

There is a trick: `extend()` rejects `array` object with different typecode, but if we take `iter()`, it accepts.

Actually, the constructor does not call these methods, it only executes an equivalent code, so you cannot change the behavior in subclasses by overriding them. But it is too low level implementation detail.

File: Doc/library/array.rst [hunk 3]
```diff
@@ -244,7 +247,7 @@ The module defines the following type:
       obtain a unicode string from an array of some other type.
 
 
-When an array object is printed or converted to a string, it is represented as
+When an array object is converted to a string, it is represented as
```
Comment (serhiy-storchaka) [inline Doc/library/array.rst]: In very old Python printing was a separate operation from converting to string.

File: Doc/library/array.rst [hunk 4]
```diff
@@ -79,14 +79,16 @@ The module defines the following type:
...
+   or :class:`bytearray` object, a unicode string, or iterable over elements
+   of the appropriate type.
 
-   If given a list or string, the initializer is passed to the new array's
-   :meth:`fromlist`, :meth:`frombytes`, or :meth:`fromunicode` method (see below)
-   to add initial items to the array.  Otherwise, the iterable initializer is
-   passed to the :meth:`extend` method.
+   If given a :class:`bytes` or :class:`bytearray` object, the initializer
+   is passed to the new array's :meth:`frombytes` method;
+   if given a unicode string, the initializer is passed to the
```
Comment (hugovk) [inline Doc/library/array.rst]:
```suggestion
   if given a Unicode string, the initializer is passed to the
```

File: Doc/library/array.rst [hunk 5]
```diff
@@ -152,10 +154,11 @@ The module defines the following type:
       must be the right type to be appended to the array.
 
 
-   .. method:: frombytes(s)
+   .. method:: frombytes(buffer)
 
-      Appends items from the string, interpreting the string as an array of machine
-      values (as if it had been read from a file using the :meth:`fromfile` method).
+      Appends items from the :term:`bytes-like object`, interpreting
+      it's content as an array of machine values (as if it had been read
```
Comment (hugovk) [inline Doc/library/array.rst]:
```suggestion
      its content as an array of machine values (as if it had been read
```

**Review by hugovk — APPROVED** [review]

**Review by hugovk — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
@hugovk, I made yet one change after your approving, could you please look at it again? The text was not clean, `tounicode()` also "converts array to string", but that paragraph was about `repr()`.

File: Doc/library/array.rst [hunk 6]
```diff
@@ -239,24 +242,27 @@ The module defines the following type:
...
-      obtain a unicode string from an array of some other type.
+      obtain a Unicode string from an array of some other type.
 
 
-When an array object is printed or converted to a string, it is represented as
-``array(typecode, initializer)``.  The *initializer* is omitted if the array is
-empty, otherwise it is a string if the *typecode* is ``'u'`` or ``'w'``,
-otherwise it is a list of numbers.
-The string is guaranteed to be able to be converted back to an
+The string representation of array objects has a form
```
Comment (hugovk) [inline Doc/library/array.rst]:
```suggestion
The string representation of array objects has the form
```

**Review by hugovk — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
Thank you for your help @hugovk. It is priceless to me, with my level of English.

**Pull request merged**